service/history/queues/action_pending_task_count.go

```diff
@@ -0,0 +1,226 @@
+			for namespaceID, pendingTaskCount := range a.pendingTaskPerNamespacePerSlice[s] {
+				a.tasksPerNamespace[namespaceID] += pendingTaskCount
+				a.slicesPerNamespace[namespaceID] = append(a.slicesPerNamespace[namespaceID], s)
+			}
+		})
+	}
+	for _, sliceList := range a.slicesPerNamespace {
+		slices.SortFunc(sliceList, func(this, that Slice) bool {
+			thisMin := this.Scope().Range.InclusiveMin
+			thatMin := that.Scope().Range.InclusiveMin
+			return thisMin.CompareTo(thatMin) > 0
+		})
+	}
+}
+
+func (a *actionQueuePendingTask) findSliceToClear(
+	currentPendingTasks int,
+	targetPendingTasks int,
+) {
+	// order namespace by # of pending tasks
+	pq := collection.NewPriorityQueue(func(this, that namespace.ID) bool {
+		return a.tasksPerNamespace[this] > a.tasksPerNamespace[that]
+	})
+	for namespaceID := range a.tasksPerNamespace {
+		pq.Add(namespaceID)
+	}
+
+	for currentPendingTasks > targetPendingTasks && !pq.IsEmpty() {
+		namespaceID := pq.Remove()
+
+		sliceList := a.slicesPerNamespace[namespaceID]
+		if len(sliceList) == 0 {
+			panic("Found namespace with non-zero pending task count but has no correspoding Slice")
+		}
+
+		// pop the first slice in the list
+		sliceToClear := sliceList[0]
+		sliceList = sliceList[1:]
+		a.slicesPerNamespace[namespaceID] = sliceList
+
+		a.tasksPerNamespace[namespaceID] -= a.pendingTaskPerNamespacePerSlice[sliceToClear][namespaceID]
+		if a.tasksPerNamespace[namespaceID] > 0 {
+			pq.Add(namespaceID)
+		}
+
+		a.namespaceToClearPerSlice[sliceToClear] = append(a.namespaceToClearPerSlice[sliceToClear], namespaceID)
+	}
+}
+
+func (a *actionQueuePendingTask) splitAndClearSlice(
+	readers map[int32]Reader,
+	readerGroup *ReaderGroup,
+) {
+	for readerID, reader := range readers {
+		if readerID == int32(a.maxReaderCount)-1 {
+			// we can't do further split, have to clear entire slice
+			cleared := false
+			reader.ClearSlices(func(s Slice) bool {
+				_, ok := a.namespaceToClearPerSlice[s]
+				cleared = cleared || ok
+				return ok
+			})
+			if cleared {
+				reader.Pause(clearSliceThrottleDuration)
+			}
+			continue
+		}
+
+		var splitSlices []Slice
+		reader.SplitSlices(func(s Slice) ([]Slice, bool) {
+			namespaceIDs, ok := a.namespaceToClearPerSlice[s]
+			if !ok {
+				return nil, false
+			}
+
+			namespaceIDStrings := make([]string, 0, len(namespaceIDs))
+			for _, namespaceID := range namespaceIDs {
+				namespaceIDStrings = append(namespaceIDStrings, namespaceID.String())
+			}
+
+			split, remain := s.SplitByPredicate(tasks.NewNamespacePredicate(namespaceIDStrings))
+			split.Clear()
+			splitSlices = append(splitSlices, split)
+			return []Slice{remain}, true
+		})
+
+		if len(splitSlices) == 0 {
+			continue
+		}
+
+		nextReader, ok := readerGroup.ReaderByID(readerID + 1)
+		if ok {
+			nextReader.MergeSlices(splitSlices...)
+		} else {
+			nextReader = readerGroup.NewReader(readerID+1, splitSlices...)
+		}
+		nextReader.Pause(clearSliceThrottleDuration)
```

Comment:
why pause here? (for 10s?)

Reply:
Because we just did an unloading, without pause those tasks may be loaded back immediately. 
Although `nextReader` will use a lower max task count so it won't trigger another unloading, but I think it will still increase the chance of another unloading. So pausing for a while and let some in memory tasks get processed first before attempting another load.